src/net_processing.cpp

```diff
@@ -2217,34 +2239,36 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
     }
 
     bool received_new_header = false;
-    const CBlockIndex *pindexLast = nullptr;
+    const CBlockIndex* pindexLast = nullptr;
     {
         LOCK(cs_main);
         CNodeState *nodestate = State(pfrom.GetId());
 
-        // If this looks like it could be a block announcement (nCount <
-        // MAX_BLOCKS_TO_ANNOUNCE), use special logic for handling headers that
-        // don't connect:
-        // - Send a getheaders message in response to try to connect the chain.
-        // - The peer can send up to MAX_UNCONNECTING_HEADERS in a row that
-        //   don't connect before giving DoS points
-        // - Once a headers message is received that is valid and does connect,
-        //   nUnconnectingHeaders gets reset back to 0.
-        if (!m_chainman.m_blockman.LookupBlockIndex(headers[0].hashPrevBlock) && nCount < MAX_BLOCKS_TO_ANNOUNCE) {
-            nodestate->nUnconnectingHeaders++;
+        if (!LookupBlockIndexForPeer(pfrom, headers.front().hashPrevBlock)) {
             m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::GETHEADERS, m_chainman.ActiveChain().GetLocator(m_chainman.m_best_header), uint256()));
```

Comment:
Since this `getheaders` message uses `m_best_header` for the locator, it's possible that before we get the response, that `m_best_header` might change to be a different chain tip (if we learn of a more work headers chain from another peer in between).

In that situation, we might get a response from our peer that would appear to be trying to observe a header that is not permitted, and thus trigger another `getheaders` message, which would be unfortunate.

I think to avoid this, we need to explicitly add the chain tip we're building a locator from to the chain tip set for a peer network whenever we send out a getheaders. (This would cover the cases where we send a getheaders based on our active chain tip as well, in the event that we reorged away from it.)

Reply:
Good point, added a call to `UpdateChainTipSet` whenever we sent out a `getheaders`. It's probably a rather rare edge case but still worth handling imo.